src/loki/matrix_action.cc

```diff
@@ -59,7 +59,7 @@ namespace loki {
 void loki_worker_t::init_matrix(Api& request) {
   // we require sources and targets
   auto& options = *request.mutable_options();
-  if (options.action() == Options::sources_to_targets) {
+  if (options.action() == Options::sources_to_targets || options.action() == Options::expansion) {
```

Comment:
just tons of this plumbing. the only real change is adding the expansion_callback member and calling that during the matrix expansion